### PR TITLE
Lesson 7 Part 3: Simple bash script for module testing

### PR DESCRIPTION
**Pull request by Noxybot**

### CODE REVIEW

File: 02-KernelModules_CharacterDevice/Makefile [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+obj-m := Character_driver.o
+
+.PHONY: all clean
+
+all:
+	$(MAKE) -C ${KDIR} M=$$PWD modules
+
+clean:
+	$(MAKE) -C ${KDIR} M=$$PWD clean
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/Makefile:11]:
```suggestion
	$(MAKE) -C ${KDIR} M=$$PWD clean
	
```

Please add empty line at the end of file

**Review by galkinletter — CHANGES_REQUESTED** [review]
Need fix

File: 02-KernelModules_CharacterDevice/test_script.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+PATH_TO_DRIVER="/home/user/char_driver/Character_driver.ko"
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/test_script.sh:3]: Need PATH like variables in makefile

**Review by galkinletter — CHANGES_REQUESTED** [review]

File: 02-KernelModules_CharacterDevice/Makefile [hunk 3]
```diff
@@ -0,0 +1,11 @@
+KDIR := /home/eduard/Kernel_i386
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/Makefile:1]: Need PATH like variables in makefile

File: 02-KernelModules_CharacterDevice/Create_device.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
+#pragma once
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/Create_device.h]: Using #pragma once should work on any modern compiler, but I don't see any reason not to use a standard #ifndef include guard. It works just fine. The one caveat is that GCC didn't support #pragma once before version 3.4.

File: 02-KernelModules_CharacterDevice/Character_driver.c [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    cdev_del(&my_dev);
+    class_destroy(my_class);
+    unregister_chrdev_region(dev, 1);
+
+    // Free allocated for driver buffer memory
+    delete_buffer();
+}
+
+module_init(char_driver_init);
+module_exit(char_driver_exit);
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/Character_driver.c]: Please, add empty line at the end of file

File: 02-KernelModules_CharacterDevice/Create_device.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    ret = len;
+    goto out;
+    }
+    
+    
+
+    out:
+        printk(KERN_DEBUG "Wrote %d bytes, my buffer size now is: %d\n", len, data_size);
+        return ret;
+}
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/Create_device.h]: Please, add empty line at the end of file

File: 02-KernelModules_CharacterDevice/Memory_management.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+
+static void delete_buffer(void){
+    if(driver_buffer != NULL){
+        kfree(driver_buffer);
+        driver_buffer = NULL;
+    }
+    data_size = 0;
+}
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/Memory_management.h]: Please, add empty line at the end of file

File: 02-KernelModules_CharacterDevice/Sys_interface.h [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    }
+    return count;
+}
+static struct  kobject *kobj_dir;
+static struct kobj_attribute file_attr = __ATTR(SYS_FILE_NAME, S_IWUSR, NULL, store_clear);
+
+void init_sys(void){
+    kobj_dir = kobject_create_and_add(SYS_DIR_NAME, NULL);
+    sysfs_create_file(kobj_dir, &file_attr.attr);
+}
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/Sys_interface.h]: Please, add empty line at the end of file

File: 02-KernelModules_CharacterDevice/test_script.sh [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+printf "Stats after buffer clearing:\n"
+
+cat /proc/My_proc_stats/Info
+
+printf "\nTrying to read data from device after clearing (should display nothing)...\n"
+
+cat /dev/My_character_device
+
+rmmod Character_driver
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/test_script.sh:49]: Please, add empty line at the end of file

**Review by galkinletter — CHANGES_REQUESTED** [review]
Need fix

File: 02-KernelModules_CharacterDevice/character_driver.c [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    cdev_del(&my_dev);
+    class_destroy(my_class);
+    unregister_chrdev_region(dev, 1);
+
+    // Free allocated for driver buffer memory
+    delete_buffer();
+}
+
+module_init(char_driver_init);
+module_exit(char_driver_exit);
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/character_driver.c:242]:
```suggestion
module_exit(char_driver_exit);

```
Please, add empty line at the end of file

File: 02-KernelModules_CharacterDevice/device_creation.h [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+#include <linux/device.h> //for device_creaete
+#include "Memory_management.h" //driver_buffer, read_position, data_length
+#define NUMBER_OF_DEVICES   1
+#define DEVICE_NAME "My_character_device"
+
+ssize_t dev_read (struct file *file_p, char __user *usr_buffer, size_t len, loff_t *offset);
+ssize_t dev_write (struct file *file_p, const char __user *usr_buffer, size_t len, loff_t *offset);
+
+
+static struct file_operations my_fops = {
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/device_creation.h]:
```suggestion

```
Remove implementation  a function . There must be a function declaration
Remove file_operation from .h

**Review by galkinletter — CHANGES_REQUESTED** [review]
Need fix 

Remove implementation. There must be a function declaration in *.h

File: 02-KernelModules_CharacterDevice/character_driver.c [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    .read = dev_read,
+    .write = dev_write
+};
+
+static dev_t dev; // major and minor number
+
+static struct cdev my_dev;
+static struct class *my_class;
+void init(void){
+    alloc_chrdev_region(&dev, 0/*statring from minor number 0*/, NUMBER_OF_DEVICES, DEVICE_NAME); //ask kernel to allocate free major number
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/character_driver.c]: alloc_chrdev_region() returns 0 or error code.

File: 02-KernelModules_CharacterDevice/character_driver.c [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    cdev_init(&my_dev, &my_fops);
+    cdev_add(&my_dev, dev, 1);
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/character_driver.c]: cdev_add() returns a negative error code on failure.

File: 02-KernelModules_CharacterDevice/character_driver.c [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+    my_class = class_create(THIS_MODULE, "Char_driver");
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/character_driver.c:128]: There are no any error handling
if(IS_ERR(my_class)) ...

**Review by oleg-h2o — CHANGES_REQUESTED** [review]
Do handle errors, please

**Pull request closed without merging**